validation/validate.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python3
+
+import argparse
+import re
+from typing import TypedDict
+from enum import Flag, auto
+from dataclasses import dataclass
+
+
+class CommandFlag(Flag):
+    GLOBAL = auto()
+    APPEND = auto()
+    FORMAT = auto()
+
+
+def flags_from_str(args: str) -> set[CommandFlag]:
+    """Returns the set of flags present in the given args."""
+    flags: set[CommandFlag] = set()
+    for c in args:
+        match c:
+            case "a":
+                flags.add(CommandFlag.APPEND)
+            case "g":
+                flags.add(CommandFlag.GLOBAL)
+            case "F":
+                flags.add(CommandFlag.FORMAT)
+    return flags
+
+
+TmuxCommand = TypedDict(
+    "TmuxCommand",
+    {
+        "command": str,
+        "option": str | None,  # If this is setting an option, the target option.
+        "flags": set[CommandFlag],  # Flags such as '-g', '-F'.
+        "value": str,
+        # Metadata
+        "line_number": int,  # The line number that this appears in in the source file.
+        "original_text": str,  # Original text of the line in the source file.
+    },
+)
+
+
+@dataclass
+class Issue:
+    """An issue with a line in a tmux config file."""
+
+    cmd: TmuxCommand
+    message: str
+
+
+def print_error(message: str, line: str, lineno: int):
+    """Prints an error that occured on the given lineno, with the contents being line."""
+    print(f"{message}, at line {lineno}: {line}")
+    exit(1)
+
+
+def parse_config(config: str, verbose: bool) -> list[TmuxCommand]:
+    commands: list[TmuxCommand] = []
+    # TODO: Handle multiline commands that contain '\'
+    for line_number, line in enumerate(config.splitlines()):
+        cmd_re = r"\s*(\w+)\s+(-\w+)?\s*(@?[\w\-_]+)?\s*[\"']?(.*)"
+        m = re.match(cmd_re, line)
```

Comment:
you can `re.compile` this pattern once at the beginning and then match with that, rather than recompiling on each iteration of the loop

Reply:
it might not be. in theory, compiling it once and then using it many times should be kinder on system resources. however, i believe python maintains its own cache for compiled regex patterns anyway, so you may see no difference at all.
i'm also not sure how long the average tmux config is; if this loop only runs a few hundred times it would be a negligible difference.
thought it was worth pointing out for consideration regardless.

Reply:
It is cached, and I'm not really interested in micro optimisations without a measurable performance benefit that users would notice

Reply:
as i said, just a suggestion :)